other/relax/relax.jl

```diff
@@ -0,0 +1,117 @@
+end
+
+function logistic_logpdf(x, mu = 0, scale = 1)
+    y = (x - mu) / (2 * scale)
+    -2 * logaddexp(y, -y) - log(scale)
+end
+
+bernoulli_sample(logit_theta, noise) = logit(noise) < logit_theta
+
+# relaxed_bernoulli_sample(logit_theta, noise, log_temperature) = bernoulli_softmax(logistic_sample(noise, logit_theta), log_temperature)
+relaxed_bernoulli_sample(logit_theta, noise, log_temperature) = bernoulli_softmax(logistic_sample(noise, expit.(logit_theta)), log_temperature)
+
+"""
+    Computes p(u|b), where b = H(z), z = logit_theta + logit(noise), p(u) = U(0, 1)
+"""
+function conditional_noise(logit_theta, samples, noise)
+    uprime = expit.(-logit_theta)
+    @. samples * (noise * (1 - uprime) + uprime) + (1 - samples) * noise * uprime
+end
+
+"""
+    log Bernoulli(targets | theta), targets are 0 or 1.
+"""
+bernoulli_logprob(logit_theta, targets) = -logaddexp(0, -logit_theta * (targets * 2 - 1))
+
+
+"""
+    function reinforce(θ, u, x)
+
+    θ, u, x =  param, noise, func_vals
+"""
+function reinforce(θ, noise, x)
+    samples = bernoulli_sample.(θ, noise)
+    x .* Flux.gradient((θ) -> sum(bernoulli_logprob.(θ, samples)), θ; nest = true)[1]
+end
+
+
+function concrete(θ, log_temperature, noise, f)
+    relaxed_samples = relaxed_bernoulli_sample.(θ, noise, log_temperature)
+    f(relaxed_samples)
+end
+
+
+function relax(θ, u, v, log_temperature, surrogate, f)
+    samples = bernoulli_sample.(θ, u)
+    ∇surrogate = Flux.gradient(θ -> sum(concrete(θ, log_temperature, u, surrogate)), θ; nest = true)[1]
+
+    cond_noise = conditional_noise.(θ, samples, v)  # z tilde
+    cond_surrogate = concrete(θ, log_temperature, cond_noise, surrogate)
+    ∇cond_surrogate = Flux.gradient(() -> sum(cond_surrogate), Params(θ); nest = true)[θ]
+
+    func_vals = f(samples)
+    (func_vals, reinforce(θ, u, func_vals .- cond_surrogate) .+ ∇surrogate .- ∇cond_surrogate)
+end
+
+"""
+
+    est_params = (τ, ϕ)
+    θ --- parameters of Bernoulli probability distribution
+    ϕ --- parameters of surrogate network
+"""
+function relax_all(θ, ϕ, τ, surrogate, u, v, f)
+    # Returns objective, gradients, and gradients of variance of gradients.
+    θm = param(repeat(Flux.data(θ), inner = (1, size(u, 2))))
+    func_vals, ∂f_∂θ = relax(θm, u, v, τ, surrogate, f)
+    ∂var_∂ϕ = Flux.gradient(() -> sum(∂f_∂θ .^ 2) ./ size(u, 2), Params(ϕ); nest = true)
+    
+    for p in ϕ
+        p.grad .= Flux.data(∂var_∂ϕ[p])
+    end
+    θ.grad .= dropdims(mean(Flux.data(∂f_∂θ), dims = 2), dims = 2)
+    func_vals
+end
+
+
+D = 100
+θ = param(zeros(D))
+c = collect(range(0,1,length = D))
+f(x) = sum((x .- c).^2, dims = 1)
+τ = param([0.0])
+surrogate = Chain(Dense(D, 5, relu), Dense(5, 1))
+ϕ = params(surrogate)
+push!(ϕ, τ)
+nsamples = 10
+
+opt = ADAM(0.1)
+for i in 1:1000
+    u, v = rand(D, nsamples), rand(D, nsamples)
+    fVal = relax_all(θ, ϕ, τ, surrogate, u, v, f)
+    Flux.Optimise.update!(opt, ϕ)
+    Flux.Optimise.update!(opt, [θ])
```

Comment:
This is knit-picky, but is it possible to write `θ` and `ϕ` in such a way that they're interfaced in the same way?

I remember this being quite annoying when I was working with LAX.

E.g. `ϕ` you have `Flux.Optimise.update!(opt, ϕ)` but `θ` needs to be wrapped in an array `Flux.Optimise.update!(opt, [θ])`.

Or above, 

```
    for p in ϕ
        p.grad .= Flux.data(∂var_∂ϕ[p])
    end
``` 

but `θ.grad .= dropdims(mean(Flux.data(∂f_∂θ), dims = 2), dims = 2)`

I don't know if this is worthwhile, but for aesthetics and to avoid confusion, it would be nice to have both of these sets of parameters handled with the same syntax.

Reply:
Frankly speaking, I do not know how to do that at the moment. Relax (and reinforce as well) are weird in the sense that they encapsulated the forward part of the function and the loss as well. Moreover, loss needs to output values on individual samples.

I do not know, how to do this from the software engineering point of view.